rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+the `domain` of a curve from `[0.0, 1.0]` to `[0.0, 2.0]` by linearly stretching it:
+```rust
+let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x + 1.0);
+let domain = my_curve.domain();
+let scaled_curve = my_curve.reparametrize(interval(0.0, 2.0).unwrap(), |t| t / 2.0);
+```
+(A convenience method `reparametrize_linear` exists for this specific kind of thing.)
+
+However, `reparametrize` can do much more than this alone. For example, here we use `reparametrize` to
+create a new curve that is a segment of the original one:
+```rust
+let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x * 2.0);
+// The segment of `my_curve` from `0.5` to `1.0`, shifted back to `0.0:
+let curve_segment = my_curve.reparametrize(interval(0.0, 0.5).unwrap(), |t| 0.5 + t);
+```
+
+And here, we use it to reverse our curve:
+```rust
+let my_curve = function_curve(interval(0.0, 2.0).unwrap(), |x| x * x);
+let domain = my_curve.domain();
+let reversed_curve = my_curve.reparametrize(domain, |t| domain.end() - t);
+```
+
+And here, we reparametrize by an easing curve:
+```rust
+let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x + 5.0);
+let easing_curve = function_curve(interval(0.0, 1.0).unwrap(), |x| x * x * x);
+let eased_curve = my_curve.reparametrize(easing_curve.domain(), |t| easing_curve.sample(t));
+```
+(The latter also has a convenience method, `reparametrize_by_curve`, which handles the domain automatically.)
+
+---
+
+Next, we have `graph`, the last of the general functional methods of the API:
+```rust
+/// Create a new [`Curve`] which is the graph of this one; that is, its output includes the
+/// parameter itself in the samples. For example, if this curve outputs `x` at time `t`, then
+/// the produced curve will produce `(t, x)` at time `t`.
+fn graph(self) -> impl Curve<(f32, T)>
+where
+    Self: Sized,
+{ //... }
+```
+This is a subtle method whose main applications involve allowing more complex things to be expressed. For example,
+here we modify a curve by making its output value attenuate with time:
+```rust
+// A curve with a value of `3.0` over the `Interval` from `0.0` to `5.0`:
+let my_curve = const_curve(interval(0.0, 5.0).unwrap(), 3.0);
+// The same curve but with exponential falloff in time:
+let new_curve = my_curve.graph().map(|(t, x)| x * (-t).exp2());
+```
+
+Most general operations that one could think up for curves can be achieved by clever combinations of `map` and
+`reparametrize`, perhaps with a little `graph` sprinkled in.
+
+### Sampling, resampling, and interpolation
+
+The next part of the API concerns itself with more imperative (i.e. data-focused) matters. 
+
+Often, one will want to define a curve using some kind of interpolation over discrete data or, conversely,
+extract lists of samples that approximate a curve, or convert a curve to one which has been discretized.
+
+For the first of these, there is the type `SampleCurve<T, I>`, whose constructor looks like this:
```

Comment:
> I'm not sure about the name of `SampleCurve<T,I>`. It does not really convey what it does. It could also be called a `PiecewiseCurve<T,I>` or `InterpolatedCurve<T,I>`. But those names come with their own issues.

Yeah, at one point I called this "SampleInterpolatedCurve", which is what I kind of want to convey, but it's also too long :P 

> Considering that `I` interpolates between two samples and without information about derivatives, the interpolation options are very limited (e.g. step, nearest, linear and perhaps some smoothstep). Because of that, I'd suggest tot have the name of `SampleCurve<T,I>` be an implementation detail and only expose specific curve types e.g. `PiecewiseConstantCurve` for `I=step`, `PiecewiseLinearCurve` for `I=lerp`.

Well, in principle these can use any information from the sample values at each point, which could make them considerably more complex. On the other hand, the purpose of making this just take an `Fn` is explicitly to move the choice of interpolation modes "upward" to consumers in general (for example, to make interpolation guided by an enum or similar).
 
> A constant curve is still a perfectly valid curve for when only 1 sample is provided. If this can be implemented with negligible overhead, this is preferred over introducing a corner case where a sensible default is available.

My thoughts on this right now are that we should probably use the number of segments instead of `samples`, which seems more intuitive to me. That also has the side-effect of leaving only the error case of zero segments, which seems kind of unavoidable regardless (if only it were ergonomic for users to provide a `NonZeroUsize`....).

> Scope creep / things that we may want to support in a future version:
> 
> > It makes sense to have a generic class for piecewise/interpolated curves as there's a lot of common code between these types of curves.

Yeah; perhaps this could be a part of the story for the analogue of `FromIterator`/`collect` at some point.

> > 2. Allow samples to include derivative information. (i.e. `samples: impl Into<Vec<Sample<T>>`, where Sample indicates what derivative information is available. Then implement `impl<T> From<T> for Sample<T>` for ease of use.)

Yeah, there is some [ongoing work](https://github.com/bevyengine/bevy/pull/14306) on including derivative information as well; I think some general form of Hermite approximation for differentiable curves would be quite useful.